File: Greedy/codeforces_1931E.py
# Problem link ===>> https://codeforces.com/problemset/problem/1931/A

def count_digits(num):
  count = 0
  while num > 0:
    num //= 10 
    count += 1
  return count


def count_zeroes(num):
  count = 0
  while num > 0 and num % 10 == 0:
    num //= 10 
    count += 1
  return count


def main():
  t = int(input())

  for _ in range(t):
    n, m = map(int, input().split())

    nums = list(map(int, input().split()))

    digit_sum = 0
    zeroes = [0] * n 

    for i in range(n):
      digit_sum += count_digits(nums[i])
      zeroes[i] = count_zeroes(nums[i])

    zeroes.sort(reverse=True)

    removed_zeroes = 0

    for i in range(0, n, 2):
      removed_zeroes += min(zeroes[i], m)
      m -= removed_zeroes 

    digit_sum -= removed_zeroes

    if digit_sum >= (m + 1):
      print("Sasha")
    else:
      print("Anna")


if __name__ == "__main__":
  main()

File: Greedy/codeforces_1931E.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/1931/A
// submission link ===>> https://codeforces.com/contest/1931/submission/246180400

#include <bits/stdc++.h>
using namespace std;

long long countDigits(long long num)
{
    long long count = 0;
    while (num > 0)
    {
        num /= 10;
        count++;
    }
    return count;
}

long long countZeroes(long long num)
{
    long long count = 0;
    while (num > 0 && num % 10 == 0)
    {
        num /= 10;
        count++;
    }
    return count;
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, m;
        cin >> n >> m;

        vector<int> nums(n);
        for (int i = 0; i < n; i++)
        {
            cin >> nums[i];
        }

        int digitSum = 0;
        vector<int> zeroes(n);
        for (int i = 0; i < n; i++)
        {
            digitSum += countDigits(nums[i]);
            zeroes[i] = countZeroes(nums[i]);
        }

        sort(zeroes.begin(), zeroes.end(), greater<int>());

        for (int i = 0; i < n; i += 2)
        {
            digitSum -= zeroes[i];
        }

        if (digitSum >= (m + 1))
        {
            cout << "Sasha" << endl;
        }
        else
        {
            cout << "Anna" << endl;
        }
    }
    return 0;
}